src/mongo_odm/query_builder.hpp

```diff
@@ -0,0 +1,473 @@
+    return ComparisonExpr<Base, T>(lhs, rhs, QS_LT);
+}
+
+template <typename Base, typename T, typename U,
+          typename = typename std::enable_if_t<!std::is_same<T, bool>::value>>
+constexpr ComparisonExpr<Base, T> operator<=(const Nvp<Base, T> &lhs, const U &rhs) {
+    return ComparisonExpr<Base, T>(lhs, rhs, QS_LTE);
+}
+
+template <typename Base, typename T,
+          typename = typename std::enable_if_t<std::is_same<T, bool>::value>>
+constexpr ComparisonExpr<Base, T> operator<=(const Nvp<Base, T> &lhs, const T &rhs) {
+    return ComparisonExpr<Base, T>(lhs, rhs, QS_LTE);
+}
+
+template <typename Base, typename T, typename U,
+          typename = typename std::enable_if_t<!std::is_same<T, bool>::value>>
+constexpr ComparisonExpr<Base, T> operator!=(const Nvp<Base, T> &lhs, const U &rhs) {
+    return ComparisonExpr<Base, T>(lhs, rhs, QS_NE);
+}
+
+template <typename Base, typename T,
+          typename = typename std::enable_if_t<std::is_same<T, bool>::value>>
+constexpr ComparisonExpr<Base, T> operator!=(const Nvp<Base, T> &lhs, const T &rhs) {
+    return ComparisonExpr<Base, T>(lhs, rhs, QS_NE);
+}
+
+/**
+ * Negates a comparison expression in a $not expression.
+ */
+template <typename Base, typename T>
+constexpr NotExpr<Base, T> operator!(const ComparisonExpr<Base, T> &expr) {
+    return NotExpr<Base, T>(expr);
+}
+
+/**
+ * Comma operator that combines two expressions or an expression and an expression list into a new
+ * expression list.
+ */
+template <typename Head, typename Tail,
+          typename = typename std::enable_if<is_expression<Head>::value &&
+                                             is_expression<Tail>::value>::type>
+constexpr ExpressionList<Head, Tail> operator,(Tail &&tail, Head &&head) {
+    return ExpressionList<Head, Tail>(head, tail);
+}
+
+/**
+ * Boolean operator overloads for expressions.
+ */
+template <typename Expr1, typename Expr2,
+          typename = typename std::enable_if<is_expression<Expr1>::value &&
+                                             is_expression<Expr2>::value>::type>
+constexpr BooleanExpr<Expr1, Expr2> operator&&(const Expr1 &lhs, const Expr2 &rhs) {
+    return BooleanExpr<Expr1, Expr2>(lhs, rhs, LO_AND);
+}
+
+template <typename Expr1, typename Expr2,
+          typename = typename std::enable_if<is_expression<Expr1>::value &&
+                                             is_expression<Expr2>::value>::type>
+constexpr BooleanExpr<Expr1, Expr2> operator||(const Expr1 &lhs, const Expr2 &rhs) {
+    return BooleanExpr<Expr1, Expr2>(lhs, rhs, LO_OR);
+}
+
+/**
+ * hasField determines whether a type Base has a member of the given type T as
+ * the Nth member out of M total members which have name value pairs.
+ */
+// By default, if N>=M the index is out of bounds and hasField is false-y.
+template <typename Base, typename T, size_t N, size_t M,
+          bool = N<M> struct hasField : public std::false_type {};
```

Comment:
for the sake of clarity, you could wrap `N<M` in parentheses so it doesn't look like N templated on M